### ISSUE TITLE
mrcat not appending dw_scheme

### ISSUE THREAD

**Issue by bjeurissen** [MRtrix3/mrtrix3#907]
Was testing the 0.3.16 branch and one of my scripts fell over on concatenating two images.

The initial images were as follows:

```$ mrinfo ${tmp}/img1.mif ${tmp}/img2.mif
************************************************
Image:               "/tmp/tmp.nDfv3FugGJ/img1.mif"
************************************************
  Dimensions:        100 x 100 x 58 x 3
  Voxel size:        2.4 x 2.4 x 2.4 x ?
  Data strides:      [ 1 2 3 4 ]
  Format:            MRtrix
  Data type:         unsigned 16 bit integer (little endian)
  Intensity scaling: offset = 0, multiplier = 1
  Transform:                    1           0           0      -122.3
                                0           1           0      -67.68
                                0           0           1      -69.52
  DW_scheme:         0,0,0,0
                     0,0,0,0
                     0,0,0,0
************************************************
Image:               "/tmp/tmp.nDfv3FugGJ/img2.mif"
************************************************
  Dimensions:        100 x 100 x 58 x 3
  Voxel size:        2.4 x 2.4 x 2.4 x ?
  Data strides:      [ 1 2 3 4 ]
  Format:            MRtrix
  Data type:         unsigned 16 bit integer (little endian)
  Intensity scaling: offset = 0, multiplier = 1
  Transform:                    1           0           0      -122.3
                                0           1           0      -67.68
                                0           0           1      -69.52
  DW_scheme:         0,0,0,0
                     0,0,0,0
                     0,0,0,0
```

And the resulting concatenate failed to append both dw_schemes:

```
$ mrcat -axis 3 ${tmp}/img1.mif ${tmp}/img2.mif - | mrinfo -
mrcat: [100%] concatenating "/tmp/tmp.nDfv3FugGJ/img1.mif"
mrcat: [100%] concatenating "/tmp/tmp.nDfv3FugGJ/img2.mif"
************************************************
Image:               "/tmp/mrtrix-tmp-QOgfFD.mif"
************************************************
  Dimensions:        100 x 100 x 58 x 6
  Voxel size:        2.4 x 2.4 x 2.4 x ?
  Data strides:      [ 1 2 3 4 ]
  Format:            Internal pipe
  Data type:         unsigned 16 bit integer (little endian)
  Intensity scaling: offset = 0, multiplier = 1
  Transform:                    1           0           0      -122.3
                                0           1           0      -67.68
                                0           0           1      -69.52
  DW_scheme:         0,0,0,0
                     0,0,0,0
                     0,0,0,0
  mrtrix_version:    0.3.15-1065-gae03932f
```

Switching to an earlier version of the stable branch, it works as expected.

```
$ mrcat -axis 3 ${tmp}/img1.mif ${tmp}/img2.mif - | mrinfo -
mrcat: [100%] concatenating "/tmp/tmp.nDfv3FugGJ/img1.mif"
mrcat: [100%] concatenating "/tmp/tmp.nDfv3FugGJ/img2.mif"
************************************************
Image:               "/tmp/mrtrix-tmp-09q8EY.mif"
************************************************
  Dimensions:        100 x 100 x 58 x 6
  Voxel size:        2.4 x 2.4 x 2.4 x ?
  Data strides:      [ 1 2 3 4 ]
  Format:            Internal pipe
  Data type:         unsigned 16 bit integer (little endian)
  Intensity scaling: offset = 0, multiplier = 1
  Transform:                    1           0           0      -122.3
                                0           1           0      -67.68
                                0           0           1      -69.52
  dw_scheme:         [ 6 entries ]
  mrtrix_version:    0.3.15-315-g18462ace
```

**Comment by bjeurissen** [comment]
Issue tracks down to:
```
mrcat [cmd/mrcat.cpp: 121]: in[n] = "/tmp/tmp.nDfv3FugGJ/img1.mif", UInt16LE, size [ 100 100 58 3 ], voxel size [ 2.4 2.4 2.4 nan ], strides [ 1 100 10000 580000 ]
mrcat [cmd/mrcat.cpp: 123]: grad = 
mrcat [cmd/mrcat.cpp: 124]: grad.rows() = 0
mrcat [cmd/mrcat.cpp: 125]: grad.cols() = 0
```

Not sure why this is happening...

**Comment by bjeurissen** [comment]
There is something in the header of these images that causes ``parse_DW_scheme`` to report an empty header:
```
mrtrix image
dim: 100,100,58,3
vox: 2.4,2.4,2.4,nan
layout: +0,+1,+2,+3
datatype: UInt16LE
transform: 1, 0, 0, -122.338
transform: 0, 1, 0, -67.68300000000001
transform: 0, 0, 1, -69.5184
DW_scheme: 0,0,0,0
DW_scheme: 0,0,0,0
DW_scheme: 0,0,0,0
mrtrix_version: 0.3.15-1105-g773e3247-dirty
file: img1.dat
```

The only thing I notice is that there is no comment field. Would this cause any issues? It didn't in earlier versions...

**Comment by jdtournier** [comment]
I have a feeling this might be to do with the `DW_scheme` header entry being case-sensitive... Should be `dw_scheme`. How were these images produced? 

Also, looking at where this might make a difference: [this](https://github.com/MRtrix3/mrtrix3/blame/tag_0.3.16/core/formats/mrtrix_utils.h#L80) might have something to do with it... @Lestropie: any particular reason to preserve the case of the header entries...?

**Comment by jdtournier** [comment]
OK, I guess the [commit message](https://github.com/MRtrix3/mrtrix3/commit/974fc448d25d977d86f0c0508974c2c52eb94e7a) makes it clear why capitalisation should be preserved. In which case we should searching for keys in a case-insensitive way in [this line](https://github.com/MRtrix3/mrtrix3/blob/tag_0.3.16/src/dwi/gradient.cpp#L76). Not sure if there's a clean way to do this...

**Comment by bjeurissen** [comment]
The file was created with the bundled matlab function ``write_mrtrix.m``, which suggests ``DW_scheme`` as the fieldname for the the gradient table.

**Comment by bjeurissen** [comment]
Could patch the matlab script to use lowercase? But other people who used it in the past will still be screwed...

**Comment by jdtournier** [comment]
I think we'll need to use [this solution](http://stackoverflow.com/a/3009806/3244166) then, making the `keyval()` entry of the header a case-insensitive `std::map`, like this:

```C++
// recommended in Meyers, Effective STL when internationalization and embedded
// NULLs aren't an issue.  Much faster than the STL or Boost lex versions.
struct lessCI : public std::binary_function<std::string, std::string, bool> {
    bool operator()(const std::string &lhs, const std::string &rhs) const {
        return strcasecmp(lhs.c_str(), rhs.c_str()) < 0 ;
    }
};

typedef std::map<std::string, std::string, lessCI> mapCI;

// in header.h:
class Header { 
  ...
  mapCI keyval_;
  ...
};
```

**Comment by bjeurissen** [comment]
Should we also change the matlab scripts to lowercase for dw_scheme?

**Comment by thijsdhollander** [comment]
Probably both solutions to be safe?  If we regard our "standard" for that entry to be `dw_scheme` in lowercase, then indeed the matlab scripts we provide ourselves should probably generate just that.  Still making the general treatment of these entires case-insensitive would be a safe step that broadens compatibility with what other people may generate.

**Comment by thijsdhollander** [comment]
I've gone ahead and changed it to lowercase in the matlab scripts in the `tag_0.3.16` branch.  The funny thing is that the read script actually already had lowercase, it was just the writing one that had `DW_scheme`.  I've also changed the matlab variable names to lowercase (even though that wouldn't matter of course), so the mistake is less likely to sneak in again if other people write scripts inspired by those.  All occurences of "`dw_scheme`" are now lowercase.

**Comment by jdtournier** [comment]
There shouldn't be any need to update the matlab scripts. It'll fix this issue, but won't fix the underlying problem. The issue really was that the mif format (as implemented) was case-insensitive (we should document that), but after the changes made [here ](https://github.com/MRtrix3/mrtrix3/blame/tag_0.3.16/core/formats/mrtrix_utils.h#L80) to support BIDS, the keys are no longer converted to lowercase when reading the header. In general, I think it's a good idea to preserve the case of the entries, but that means we now need to change things to allow case-insensitive retrieval of header entries in the `keyval` dictionary. That's what the changes I mentioned above relate to.

**Comment by Lestropie** [comment]
Closing as hopefully addressed in #1555; but also initial report was back on `0.3.16`.
